### ISSUE TITLE
Feature Request : Have cider-eval-last-sexp prompt for context

### ISSUE THREAD

**Issue by l0st3d** [clojure-emacs/cider#2113]
It might be nice if, when trying to eval an sexp from the middle of a function, there was some way to inject some form of context. I'm imagining something like wrapping the form in a let before trying to eval it. Ideally CIDER would prompt me for each of the undefined forms either by analysing the code before evaluation, and remember my previous input, but I'd be happy with just being prompted for a let-binding form to insert, which could default to an empty vector. The idea is to not have to pollute the namespace I'm working in with temporary vars that I need to remember to remove and undefine later.

**Comment by xiongtx** [comment]
I don't think that's a good idea. `cider-eval-last-sexp` should do what it says--eval the last sexp. If you want a `let`-binding, include it as part of the sexp you want to eval.

**Comment by ChrisBlom** [comment]
I thinks its a good idea, it would be useful to quickly test a sexp without needing to adding temporary changes to the buffer. Maybe not as default behaviour for `cider-eval-last-sexp` though, but as a separate command or with a prefix key.

**Comment by bbatsov** [comment]
That should definitely not be the default behaviour. 

The debugger already has similar functionality, so I'm guessing it can be leveraged for this as well. Whether this is easy or not I do not know - maybe @Malabarba can shed some light on this.

**Comment by jpmonettas** [comment]
@l0st3d have you tried https://github.com/vvvvalvalval/scope-capture ?

**Comment by bbatsov** [comment]
Btw, I know it's not the same, but some of you might find useful the new `cider-eval-defun-to-point`. It runs the current top-level expression up to the cursor's position, which is something like a contextual evaluation (depending on your perspective).

**Comment by bbatsov** [comment]
Reading again @l0st3d's simple idea about just asking the user to supply some let-bindings and remembering those across invocations I decided to implement it. Parsing an expression for locals and prompting just for those would be fancy and is not a rocket science, but I simply don't have time to work on it, so let's just leave it to the user to provide whatever bindings they want.

**Comment by viesti** [comment]
I just found this cool feature! :) I ran into it while looking if there would be a way to define a dynamic var around a form in the middle of a clojure.test test.

I've been writing tests so that I have a `fixture/*test-system*` dynamically bound to an integrant or component system and then I run the test code against a live system at the repl too. Now I've been copying the test form into repl, wrapped in with:

```clojure
(binding [fixture/*test-system* reloaded.repl/system]
  (is 42 (get-answer-from-api)))
``` 

`cider-eval-last-sexp-in-context` is eerily close to this idea 😅
